Constrain generic type on NumericAssertions

jnyrup opened the pull request:
Further constraining of `T` to `IComparable<T>` gets us:
* No more runtime type check in constructor
* Avoid (un)boxing of primitives
* Type preserving `Subject`, i.e. `T`/`T?` instead of `IComparable<T>`

lg2de commented on Src/FluentAssertions/Numeric/NumericAssertions.cs:
```diff
@@ -14,21 +14,21 @@ namespace FluentAssertions.Numeric
         }
 
-        public IComparable Subject { get; }
+        private protected NumericAssertions(T? value)
```
Why `private protected`?
Do you want forbid extensions using this class?

jnyrup replied:
I don't think I'm forbidding extensions on this class.
Did you see the other constructors?
There are two constructor available, which I think should be covering
```c#
public NumericAssertions(T value) { }
public NullableNumericAssertions(T? value) { }
```

jnyrup commented:
Glad to hear there was a reason back then.
The somewhat odd construction have stricken me every time I've seen it over the past years.

Tried to dig in history to see if I could find a cause:

1eb8f4dfec1f848148bc05e73183e4979ef43888 
* Removed `T : struct, Comparable` constraints (I'm noticing the non-generic `IComparable`)
* Changed constructor from `T?` to `T`
* Added runtime `IComparable` check
* Changed `Subject` from `T?` to `IComparable`

bddb0247760eec7440116a08bf79169529121e50
* Added `T : struct`
* Changed constructor from `T` to `object`

But I didn't find any compilation issues...
